angel_system/impls/detect_activities/two_stage/two_stage_detect_activities.py

```diff
@@ -0,0 +1,186 @@
+        self._crop_size = 224
+        self._frames_per_second = 30
+
+        # Pytorch default transfroms for imagenet pre-trained models
+        self.transform = torchvision.transforms.Compose(
+            [
+                torchvision.transforms.ToTensor(),
+                torchvision.transforms.Normalize(
+                    self._mean, self._std
+                )
+            ]
+        )
+
+        # TODO: Set up the labels from the given labels file
+        self._labels = H2O_CLASSES
+        # with open(self._labels_file, "r") as f:
+        #     for line in f:
+        #         self._labels.append(line.rstrip())
+
+    def get_model(self) -> torch.nn.Module:
+        """
+        Lazy load the torch model in an idempotent manner.
+        :raises RuntimeError: Use of CUDA was requested but is not available.
+        """
+        model = self._model
+        if model is None:
+            # Load the model with the checkpoint
+            model = TwoStageModule(self._checkpoint_path, self._num_classes)
+            model = model.eval()
+
+            # Transfer the model to the requested device
+            if self._torch_device != 'cpu':
+                if torch.cuda.is_available():
+                    model_device = torch.device(device=self._torch_device)
+                    model = model.to(device=model_device)
+                else:
+                    raise RuntimeError(
+                        "Use of CUDA requested but not available."
+                    )
+            else:
+                model_device = torch.device(self._torch_device)
+
+            self._model = model
+            self._model_device = model_device
+
+        return model
+
+    def detect_activities(
+        self,
+        frame_iter: Iterable[np.ndarray],
+        aux_data_iter: Iterable[Dict]
+    ) -> Iterable[str]:
+        """
+        Formats the given iterable of frames into the required input format
+        for the swin model and then inputs them to the model for inferencing.
+        """
+        # Check that we got the right number of frames
+        frame_iter = list(frame_iter)
+        aux_data = dict(aux_data_iter)
+        assert all([len(frame_iter) == len(aux_data[i]) for i in aux_data])
+        # assert len(frame_iter) == (self._sampling_rate * self._num_frames)
+        model = self.get_model()
+
+        # Apply data pre-processing
+        frames = [self.transform(f) for f in frame_iter]
+        frames = torch.stack(frames)
+        for k in aux_data:
+            aux_data[k] = torch.Tensor(aux_data[k])
+
+        # Move the inputs to the GPU if necessary
+        if self._model.cuda:
+            frames = frames.cuda()
+            for k in aux_data:
+                aux_data[k] = aux_data[k].cuda()
+
+        # Predict!
+        with torch.no_grad():
+            preds = self._model(frames, aux_data)
+
+        # Get the top predicted classes
+        post_act = torch.nn.Softmax(dim=1)
```

Comment:
Could this be created once in the constructor?

Reply:
Yes, we can. But I removed class and am directly using the softmax functional from pytorchnn.functional (as `post_act` is not used anywhere else).